applications/ConstitutiveLawsApplication/custom_constitutive/traction_separation_law.cpp

```diff
@@ -0,0 +1,751 @@
+TractionSeparationLaw3D<TDim>::~TractionSeparationLaw3D()
+{
+};
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+bool TractionSeparationLaw3D<TDim>::Has(const Variable<Vector>& rThisVariable)
+{
+    auto p_constitutive_law_vector = this->GetConstitutiveLaws(); 
+    bool has = false;
+    
+    if (rThisVariable == DELAMINATION_DAMAGE_VECTOR_MODE_ONE) {
+        has = true;
+    } else if (rThisVariable == DELAMINATION_DAMAGE_VECTOR_MODE_TWO) {
+        has = true;
+    } else {
+        BaseType::Has(rThisVariable);
+    }
+    
+    return has;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+Vector& TractionSeparationLaw3D<TDim>::GetValue(
+    const Variable<Vector>& rThisVariable,
+    Vector& rValue
+    )
+{
+    auto p_constitutive_law_vector = this->GetConstitutiveLaws(); 
+    auto combination_factors = this->GetCombinationFactors(); 
+   
+    rValue.clear();
+
+    if (rThisVariable == DELAMINATION_DAMAGE_VECTOR_MODE_ONE) {
+        
+        rValue.resize(combination_factors.size()+1, false);
+        
+        noalias(rValue) = mDelaminationDamageModeOne;
+        return rValue;
+    } else if (rThisVariable == DELAMINATION_DAMAGE_VECTOR_MODE_TWO) {
+        
+        rValue.resize(combination_factors.size()+1, false);
+        
+        noalias(rValue) = mDelaminationDamageModeTwo;
+        return rValue;
+    } else {
+        BaseType::GetValue(rThisVariable,rValue);
+    }
+    
+    return rValue;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+bool TractionSeparationLaw3D<TDim>::ValidateInput(const Properties& rMaterialProperties)
+{
+    // We check it layer by layer
+    bool valid_input = true;
+    auto p_constitutive_law_vector = this->GetConstitutiveLaws(); 
+    for (IndexType i_layer = 0; i_layer < p_constitutive_law_vector.size(); ++i_layer) {
+        ConstitutiveLaw::Pointer p_law = p_constitutive_law_vector[i_layer];
+        Properties& r_prop = *(rMaterialProperties.GetSubProperties().begin() + i_layer);
+        if (p_law->ValidateInput(r_prop)) {
+            valid_input = false;
+            break;
+        }
+    }
+
+    return valid_input;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+ConstitutiveLaw::StrainMeasure TractionSeparationLaw3D<TDim>::GetStrainMeasure()
+{
+    auto p_constitutive_law_vector = this->GetConstitutiveLaws(); 
+    // We return the first one
```

Comment:
it is the as in the BaseType, please remove the duplicated method

Reply:
Done!